### PR TITLE
feat: Ray integration

### PR DESCRIPTION
**Pull request by jiashenC**
* Auto-detect availability of GPUs.
* Enable CircleCI auto-testing for Ray execution.
  * Only test CPU Ray on CircleCI.
  * Disable coverage report where it causes long hang for multi-process execution. 
* Address integration test failures for Ray. Here are a list of causes that I address in this PR.
  * Bug in Ray optimizer that does not correctly yield, which often results infinite loop in optimization step.
  * A tree-like execution plan was not previously supported for Ray, mainly because the queue population does not support that yet. For a quick merge now, I do a post-processing to disable Exchange plan for all execution plans that have branch.

### CODE REVIEW

**Review by gaurav274 — COMMENTED** [review]
Can you enable `circleci` to run with `ray`? It is fine if it does not pass. We can do for `python 3.10` version only.

**Comment by jiashenC** [comment]
> Can you enable `circleci` to run with `ray`? It is fine if it does not pass. We can do for `python 3.10` version only.

@gaurav274 I enable the circleci. It does not have GPU resource, so we need to change that as well.

**Comment by xzdandy** [comment]
I think lateral join becomes linear now. So ray should be able to support function scan without worrying about tree-like execution plan.

File: eva/experimental/ray/optimizer/rules/rules.py
Comment (xzdandy) [inline eva/experimental/ray/optimizer/rules/rules.py:107]: If there are no UDFs, we probably don't need ray. How about move this logic to the `check` function?

File: eva/experimental/ray/executor/exchange_executor.py
Comment (xzdandy) [inline eva/experimental/ray/executor/exchange_executor.py:75]: I feel this does not work with tree structure. Consider the following:
```mermaid
graph TD;
    R[EXCHANGE] --> A;
    A[JOIN]-->B[EXCHANGE];
    A-->C[EXCHANGE];
    B --> D[SOME UDF1];
    C --> E[SOME UDF2];
```
We never get to the right subtree.

Comment (jiashenC) [reply eva/experimental/ray/executor/exchange_executor.py:75]: I see. I will need to cover this case then. I mainly attempted to fix the following case based on our test cases. 
```mermaid
graph TD;
E1[Exchange]-->P
P[Project]-->N
N[NestedLoopJoin]-->SL
N-->SR
SL[SeqScan]-->E2
E2[Exchange]-->S1[Storage]
SR[SeqScan]-->E3
E3[Exchange]-->S2[Storage]
```

I think we can revisit the Exchange optimizer now to decide when it is actually needed and we can modify the executor based on that.

Comment (xzdandy) [reply eva/experimental/ray/executor/exchange_executor.py:76]: Got this will work. I think what we want to parallel is, for example, the UDF in the seqscan's projection. Parallelizing the Storage is not really needed. It was a "lazy" implementation, so we don't need to support partition in the Storage. But it is not  optimal.

File: eva/experimental/ray/executor/ray_stage.py
Comment (xzdandy) [inline eva/experimental/ray/executor/ray_stage.py:36]: Do need to handle `ExecutorError` in  `QueueReaderExecutor`? For example, for `StageCompleteSignal`, it needs to be put back into the input queue, so that other readers can also be notified.

File: eva/executor/execution_context.py
Comment (gaurav274) [inline eva/executor/execution_context.py:53]: we should switch to the  standard env name `CUDA_VISIBLE_DEVICES`

File: eva/experimental/ray/executor/ray_stage.py
Comment (jiashenC) [reply eva/experimental/ray/executor/ray_stage.py:36]: You are right. It needs to be populated to other queues as well.

File: eva/experimental/ray/executor/exchange_executor.py
Comment (jiashenC) [reply eva/experimental/ray/executor/exchange_executor.py:76]: I think `Storage` will still benefit from `Exchange(dop=1)` for async IOs. Otherwise, the disk read needs to happen in a blocking fashion.

My main point here is that we probably have two cases to cover for tree-like execution.
1. There exists multiple children and children are not `Exchange`. I think my current change can handle this case.
2. As you pointed out, a tree can have multiple children, which are `Exchange`. Then, we need add some mechanism for `ExchangeExecutor` to handle this case.

Maybe we have some other cases as well?

File: eva/experimental/ray/optimizer/rules/rules.py
Comment (jiashenC) [reply eva/experimental/ray/optimizer/rules/rules.py:111]: I think we have to keep the logic here because it also needs to handle the sequential `LogicalGet` to `SeqScan` case. https://github.com/georgia-tech-db/eva/blob/1aac5cf09c99e992038cd8b4091332e233a48b2d/eva/optimizer/rules/rules_manager.py#L104

File: eva/experimental/ray/executor/exchange_executor.py
Comment (xzdandy) [reply eva/experimental/ray/executor/exchange_executor.py:76]: BTW, join does not work with ray with `parallelism > 1` as for now, because two ray actors will not have the complete information. Maybe tree structure is not a priority to worry about now.

File: eva/experimental/ray/optimizer/rules/rules.py
Comment (xzdandy) [reply eva/experimental/ray/optimizer/rules/rules.py:111]: oh I see. I thought SequentialLogicalGetToSeqScan and DistributedLogicalGetToSeqScan are two different rules.

File: eva/experimental/ray/executor/ray_stage.py
Comment (xzdandy) [inline eva/experimental/ray/executor/ray_stage.py:36]: Consider the following set up
```mermaid
graph TD;
      A[Projection with parallelism=1]--> B[UDF with parallelism = 2];
      B --> C[Get with parallelism=1];
```
, and there is ExecutorError raised at the bottom Get operator. 
If I understand correctly, what will happen is that one ExecutorError will be put into the queue between UDF and Get. And one Actor will get the ExecutorError and due to the fact that it is not root, the [code](https://github.com/georgia-tech-db/eva/blob/9aa9fac520a264dfc3e26583a59d816b8bf53c8d/eva/experimental/ray/executor/exchange_executor.py#L97) will not come into play. So it will try to process on the ExecutorError object, which is not expected. Furthermore, the second actor will not be affected, because it does not get the ExecutorError, which has already been read by the first actor. I think you want to modify the [QueueReaderExecutor](https://github.com/georgia-tech-db/eva/blob/9aa9fac520a264dfc3e26583a59d816b8bf53c8d/eva/experimental/ray/executor/exchange_executor.py#L44) instead.

File: eva/optimizer/plan_generator.py
Comment (gaurav274) [inline eva/optimizer/plan_generator.py:123]: I think this code piece can be simplified. How about you assert on `is_top` and `len(plan.children) == 1` and always return `plan.children`.  This is also simplify the above check on `List`.

File: eva/experimental/ray/optimizer/rules/rules.py
Comment (gaurav274) [inline eva/experimental/ray/optimizer/rules/rules.py:116]: @kaushikravichandran This can be moved outside to a heuristic in a future PR.

**Review by gaurav274 — COMMENTED** [review]
LGTM. We can wait for Andy's review before merging.

**Comment by gaurav274** [comment]
Please add a test case where the `exchange` removal code is triggered. Thanks!

**Comment by gaurav274** [comment]
@kaushikravichandran Let us wait for a few builds and then promote `ray` from experimental to default feature. Thanks!

File: test/integration_tests/test_error_handling_with_ray.py
Comment (xzdandy) [inline test/integration_tests/test_error_handling_with_ray.py:58]: Thanks for the testcase!

File: test/optimizer/rules/test_rules.py
Comment (xzdandy) [inline test/optimizer/rules/test_rules.py:230]: Could we add a comment that indicates right now `DistributedLogicalProjectToPhysical()` and `SequentialLogicalProjectToPhysical()` are the same?

File: test/optimizer/rules/test_rules.py
Comment (xzdandy) [inline test/optimizer/rules/test_rules.py:216]: Could we add a comment that indicates `DistributedLogicalGetToSeqScan()` and `SequentialLogicalGetToSeqScan()` are same with a brief justification?

File: eva/optimizer/plan_generator.py
Comment (xzdandy) [reply eva/optimizer/plan_generator.py:123]: Is this a temporary work around? I feel We should be able to support branch and createMat. But I agree it is not priority.

**Review by xzdandy — APPROVED** [review]
Overall LGTM. Please check minor comments.

**Comment by jiashenC** [comment]
@xzdandy @gaurav274 Fix all comments. Should be ready.

**Pull request merged**
